### PR TITLE
Stop catching TypeError in groupby methods

### PR DESCRIPTION
**Pull request by jbrockmendel**
cc @WillAyd @jreback 

Note this doesn't touch cython_agg_blocks, I've run out of things to meddle with there.

Open to suggestions for the DecimalArray case.

### CODE REVIEW

File: pandas/core/groupby/ops.py [hunk 1]
```diff
@@ -526,7 +526,13 @@ def _cython_operation(self, kind, values, how, axis, min_count=-1, **kwargs):
             func = self._get_cython_function(kind, how, values, is_numeric)
         except NotImplementedError:
             if is_numeric:
-                values = ensure_float64(values)
+                try:
+                    values = ensure_float64(values)
+                except TypeError:
+                    if lib.infer_dtype(values, skipna=False) == "complex":
```
Comment (jreback) [inline pandas/core/groupby/ops.py]: future PR should push this to pandas/core/dtypes/cast.py and just call here (maybe make ensure_float64_or_complex)

**Review by jreback — CHANGES_REQUESTED** [review]
don't we need to have tests for the complex path ?

File: pandas/core/groupby/ops.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/groupby/ops.py]: yah.  i think there was some discussion of a one-pass variant of lib.infer_dtype that would go well with that

**Comment by jbrockmendel** [comment]
> don't we need to have tests for the complex path ?

we have a couple of tests that get here with an ndarray[object] containing complex objects

File: pandas/core/groupby/ops.py (hunk 1 above)
Comment (jreback) [reply pandas/core/groupby/ops.py]: great

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jbrockmendel

**Pull request merged**